### PR TITLE
[query] Fix `str.matches` to work with expressions

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 1]
```diff
@@ -2688,15 +2688,15 @@ def matches(self, regex):
...
+        regex: :class:`.StringExpression`
             Pattern to match.
 
         Returns
         -------
         :class:`.BooleanExpression`
             ``True`` if the string contains any match for the regex, otherwise ``False``.
         """
-        return to_expr(regex, tstr)._method("~", tbool, self)
+        return regex._method("~", tbool, self)
```
Comment (konradjk) [inline hail/python/hail/expr/expressions/typed_expressions.py]: and probably revert this, no?

**Comment by tpoterba** [comment]
expr_str will coerce normal str to expr.

**Comment by tpoterba** [comment]
the typecheckers are magic. they change the arguments

**Comment by konradjk** [comment]
that's cool - though worth putting the option in the docstring i think no?

**Comment by tpoterba** [comment]
I have literally no idea how to document this stuff. "it expects a string expression but there is an implicit conversion from str to StringExpression k thx bye"

**Comment by tpoterba** [comment]
we don't document both types elsewhere. Also nobody reads the docs, they just look at examples 🤷‍♂

**Comment by konradjk** [comment]
you do suggest both below in `StringExpression.replace`?

**Comment by johnc1231** [comment]
We frequently write stuff like "`str` or `StringExpression`"

**Comment by tpoterba** [comment]
ah. seems we do this somewhat frequently for the expression methods. While looking at `functions.py`, we basically never document the python objects

**Pull request merged**